matrix_sdk/src/client.rs

```diff
@@ -485,23 +485,25 @@ impl RequestConfig {
 
 impl Client {
     /// Creates a new client for making HTTP requests to the given homeserver.
+    /// Attempts to check `well_known` during the construction.
     ///
     /// # Arguments
     ///
     /// * `homeserver_url` - The homeserver that the client should connect to.
-    pub fn new(homeserver_url: Url) -> Result<Self> {
+    pub async fn new(homeserver_url: Url) -> Result<Self> {
```

Comment:
I don't think this is what was suggested in #219:

> As for doing the lookup before one is logged in, or in other words creating a Client object with only a user_id a separate async constructor that does a well-known lookup should work, no?

Please take a look at the spec as well: https://spec.matrix.org/unstable/client-server-api/#well-known-uri

This describes how and when well-known discovery should work:
1. Extract the server name from the user's Matrix ID by splitting the Matrix ID at the first colon.
2. Extract the hostname from the server name.
3. Make a GET request to https://hostname/.well-known/matrix/client.

So we'll want a new constructor pair:

```rust
pub async fn new_with_discovery(user_id: UserId) -> Result<Self> {
    // Do the well-known request here
    // If it succeeds try to query the server version to check if it's a
    // valid homeserver
}
```

If you don't like that name, perhaps `pub async fn new_from_user_id(user_id: UserId)` is better.

Reply:
Yup, makes sense, I'll do that. As a bonus, the PR should get much smaller I guess.

Reply:
Should be done now, although unfortunately I have no idea how to test it – the `mockito` server address does not allow me to create a valid user id. :/ I guess some very fancy proxying setup in the tests could fix this, but that seems like lots of work for little gain. If it's any consolation, I manually checked that it works on my homeserver.

Reply:
Create a valid user id how/where?

Reply:
I wanted to create a unit test to check whether this correctly queries and interprets the contents of `$SERVER/.well-known/matrix/client`. The constructor takes `SERVER` from the second part of the `UserId`, so creating a `UserId` with the `mockito` address as the server should work in principle. Unfortunately that adress is invalid for the purpose of creating `UserId`s.

Reply:
How did you try to construct the user id exactly?

This seems to work

```rust
let server_url = mockito::server_url();
let server_url = server_url.strip_prefix("http://").unwrap();
let user = UserId::try_from(format!("@example:{}", server_url,)).unwrap();
```

Reply:
Ouch, I am just an idiot, I forgot to remove the prefix. >.<

Should be OK and properly unit tested now.